revocation/revocation.go

```diff
@@ -0,0 +1,122 @@
+}
+
+// revocation is an internal struct used for revocation checking
+type revocation struct {
+	httpClient      *http.Client
+	mergeErrorsFunc func(errors []error) error
+}
+
+// New constructs a revocation object and substitutes default values for any that are passed as nil
+func New(httpClient *http.Client) Revocation {
+	if httpClient == nil {
+		httpClient = http.DefaultClient
+	}
+	return &revocation{
+		httpClient:      httpClient,
+		mergeErrorsFunc: DefaultMergeErrors,
+	}
+}
+
+// Validate checks OCSP, returns nil if all certs in the chain are not revoked.
+// If there is an error, it will return one of the errors defined in the revocation/ocsp package in errors.go.
+// (e.g. if a certificate in the chain is revoked by OCSP and there are no other errors, it will return ocsp.RevokedError)
+//
+// TODO: add CRL support
+func (r *revocation) Validate(certChain []*x509.Certificate, signingTime time.Time) error {
+	return r.OCSPStatus(certChain, signingTime)
+	// TODO: add CRL support
+	// https://github.com/notaryproject/notation-core-go/issues/125
+}
+
+// OCSPStatus checks OCSP, returns nil if all certs in the chain are not revoked.
+// If there is an error, it will return one of the errors defined in the revocation/ocsp package in errors.go.
+// (e.g. if a certificate in the chain is revoked by OCSP and there are no other errors, it will return ocsp.RevokedError)
+func (r *revocation) OCSPStatus(certChain []*x509.Certificate, signingTime time.Time) error {
+	return ocsp.OCSPStatus(ocsp.Options{
+		CertChain:       certChain,
+		SigningTime:     signingTime,
+		HTTPClient:      r.httpClient,
+		MergeErrorsFunc: r.mergeErrorsFunc,
+	})
+}
+
+// SetMergeErrorsFunction allows you to specify an alternative function to merge errors if the default does not fit your use case. You can also pass nil to reset it to the DefaultMergeErrors function
+func (r *revocation) SetMergeErrorsFunction(mergeErrorsFunc func(errors []error) error) {
+	if mergeErrorsFunc == nil {
+		r.mergeErrorsFunc = DefaultMergeErrors
+	} else {
+		r.mergeErrorsFunc = mergeErrorsFunc
+	}
+}
+
+// DefaultMergeErrors condenses errors for a list of errors (either for cert chain or OCSP servers) into one primary error
+func DefaultMergeErrors(errorList []error) error {
+	var result error
+	if len(errorList) > 0 {
+		result = errorList[0]
+
+		for _, err := range errorList {
+			if err == nil {
+				continue
+			}
+			switch t := err.(type) {
+			case ocsp.RevokedError:
+				// There is a revoked certificate
+				// return since any cert being revoked means leaf is revoked
+				return t
+			case ocsp.CheckOCSPError:
+				// There is an error checking
+				// return since any cert having error means chain has error (return earliest)
+				return t
+			case ocsp.UnknownStatusError:
+				// A cert in the chain has status unknown
+				// will not return immediately (in case one is revoked or has error), but will override other chain errors
+				result = t
```

Comment:
```suggestion
			case ocsp.RevokedError, csp.CheckOCSPError, ocsp.UnknownStatusError:
				result = t
```

Reply:
The issue with this is that if the last Cert in the chain has a status of Unknown, whereas a previous one had status Revoked, the cert chain would be considered Unknown rather than Revoked. My thought is that RevokedErrors and CheckOCSPErrors should be a one-and-done result for the entire chain. I can update this switch to combine RevokedErrors and CheckOCSPErrors to both return, but I think having the extra context for why is useful.

Reply:
Isn't the change proposed equivalent to your change? Am I missing something?

Reply:
The ResultFromErrors function does this a bit differently. It is similar, but a RevokedError causes it to return Revoked immediately rather than just set the current result and continue (that way other errors/statuses that are considered Unknown won't overwrite it)